doc/design/proposal-pre-import.md

```diff
@@ -0,0 +1,97 @@
+# preImport Proposal
+
+## Problem Statement
+
+With the core resolver now synchronous only, any asynchronous resolution work
+is no longer possible as the resolve function must return an immediate
+resolution.
+
+For loaders that asynchronously obtain resolution information, it would be
+useful to have a new hook to enable these use cases.
+
+## preImport Hook
+
+The hook has the signature:
+
+```ts
+export async function preImport(specifier: string, context: {
+  conditions: string[],
+  topLevel: boolean,
+  parentURL: string | undefined
+});
+```
+
+The `preImport` hook allows for tracking and asynchronous setup work for every
+top-level import operation. It has no return value, although it can return
+a promise to to delay the module pipeline operations. No further hooks will
+be called on that module graph until this hook resolves successfully if present.
+
+The `preImport` hook is called for each top-level import operation by the module
+loader, both for the host-called imports (ie for the main entry) and for dynamic
+`import()` imports. These are distinguished by the `dynamic` context.
+
+All `preImport` hooks for all loaders are run asynchronously in parallel, and
+block any further load operations (ie resolve and load) for the module graph
+being imported until they all complete successfully.
+
+Multiple import calls to the same import specifier will re-call the hook
+multiple times. The first error thrown by the `preImport` hooks will be directly
+returned to the specific import operation as the load failure.
+
+## Example
+
+<details>
+<summary>Import Map Generation</summary>
+
+Consider an import map loader which obtains the import map for a module
+asynchronously:
+
+```ts
+import { Generator } from '@jspm/generator';
+
+// stateful host import map for current session
+let importMap = { imports: {}, scopes: {} };
+
+function isUrl (specifier) {
+  try {
+    new URL(specifier);
+    return true;
+  }
+  catch {
+    return false;
+  }
+}
+
+const isBareSpecifier = id => !id.startsWith('./') && !id.startsWith('../') &&
+    !id.startsWith('/') && !isUrl(id);
+
+export async function preImport(specifier, { conditions, parentURL }) {
+  if (!isBareSpecifier(specifier)) return;
+  const generator = new Generator({
+    // passing the original map extends it
+    inputMap: importMap,
+    baseUrl: parentURL,
+    defaultProvider: 'nodemodules',
+    env: conditions,
+  });
+  await generator.install(specifier);
+  // the new map will then have the new mappings and the old mappings
+  importMap = generator.getMap();
```

Comment:
Note that this is a race condition; old `importMap` is provided to potentially multiple `Generator`s (as called out in the description: all run in parallel), then `importMap` is overridden with the results of each as they finish, so some results will get clobbered.

I'm not so concerned about this specific example, but more that this demonstrates how easy it is to make this sort of mistake with the API as proposed (parallel execution combined with forcing the output to be via global state is likely to cause many issues like this)